sof: drc: Add DRC component with floating-point calculations

johnylin76 opened the pull request:
This commit adds Dynamic Range Compression (DRC) to the list of SOF
components. DRC in audio processing is intentional to reduce the
volume of loud sounds as compressing an audio signal's dynamic range.

This is still WIP.

lgirdwood commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,849 @@
+
+	/* Q12.20 -> Q24.40 */
+	num = Q_SHIFT_LEFT((int64_t)(Q_CONVERT_FLOAT(1, 20) - knee_expf(Q_MULTSR_32X32((int64_t)(-k), (x - dk->linear_threshold), 12, 20, 20))), 20, 32);
+	return state->linear_threshold + (int32_t)(num / k);
```
out of interest, will k ever be a power of 2 ? (since shifting is always cheaper than 64bit division)

cujomalainey replied:
Also if its harded maybe we shouldnt be passing it around as an argument

johnylin76 replied:
K is not hardcoded to 5, K is the pre-calculated value determined by the compression ratio.
Then I think the division is unavoidable.

lgirdwood replied:
Is this pre-calculated based on coefficients ? Can the coefficients were generated for a k = power of 2, then it turn into a shift ?

lgirdwood reviewed (commented):
LGTM. @singalsu any comments.

cujomalainey commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,849 @@
+		drc_free_parameters(&cd->config);
+		cd->config = cd->config_new;
+		cd->config_new = NULL;
+		ret = drc_setup(cd, sourceb->stream.channels, sourceb->stream.rate);
```
i think we may have a memory leak here if we init the pre delay buffers mid stream, and im not sure if we want to reinit them. @singalsu thoughts?

johnylin76 replied:
It should not have memory leak here. In drc_setup(), drc_reset_state() is called first which freed all pre-delay buffers.

cujomalainey replied:
setup

cujomalainey commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,849 @@
+	state->ratio_base = Q_CONVERT_FLOAT(Q_CONVERT_QTOF(y0, 20) * powf(Q_CONVERT_QTOF(state->knee_threshold, 20), Q_CONVERT_QTOF(-state->slope, 20)), 30);
+}
+
+static inline void drc_set_parameters(struct drc_state *state,
```
this looks really heavy and could cause glitches, is this done once per config? if so we should do this on the cpu and pass the calculated coefficients

johnylin76 replied:
This makes senses. After more discussion and consideration, I would like to re-organize the DRC component with the following steps:
1. Because drc_set_parameters are heavy and done once per config, we should do this on the cpu and pass the calculated coefficients from sof_drc_config (coefficients are not required to be copied).
2. Implement DRC frame processing code in drc_generic.c
3. drc_state should only stores register-like variables which may be initialized by new stream, loaded and saved between frame or division processing.

lgirdwood replied:
@singalsu any inputs ?

cujomalainey replied:
@lgirdwood this is how all the processing components work including eq. I think we are good here.

johnylin76 commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,849 @@
+
+	state->db_knee = DRC_MINUS_ONE_Q24;
+	state->knee_threshold = DRC_MINUS_ONE_Q30;
+	state->ratio_base = DRC_MINUS_ONE_Q30;
```
the initialization just follows dk_init() in CRAS. In fact I suppose all state parameters should have assigned the value on drc_set_parameters(), so maybe these initialization values are not that meaningful...

johnylin76 commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,849 @@
+/* Input: Q12.20 -> Output: Q12.20 */
+static inline int32_t knee_expf(int32_t x)
+{
+#ifdef DRC_SLOW_MATHEMATICS
```
According to drc_math.h code: https://source.corp.google.com/chromeos_public/src/third_party/adhd/cras/src/dsp/drc_math.h;l=30
It chooses fast mathematics as default.

johnylin76 commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,849 @@
+	 *	 beta = -expf(k * dk->linear_threshold) / k
+	 *	 gamma = -k * x
+	 */
+	return state->knee_alpha + Q_MULTSR_32X32((int64_t)state->knee_beta, knee_expf(Q_MULTSR_32X32((int64_t)-dk->K, x, 12, 20, 20)), 30, 20, 20);
```
It is a good idea. However, after more discussion we have observed these complicated calculations are only used to determine the bunch of intermediate parameters (e.g. static curve parameters) when the input parameter set is changed. For this reason, we should put these calculations to cpu instead of the component inside. We don't have to beautify these calculations in this stage.

johnylin76 commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,849 @@
+	 * => y/x = dk->ratio_base * x^(s - 1)
+	 * => y/x = dk->ratio_base * e^(log(x) * (s - 1))
+	 */
+	return Q_MULTSR_32X32((int64_t)state->ratio_base, knee_expf(Q_MULTSR32X32((int64_t)Q_CONVERT_FLOAT(logf(Q_CONVERT_QTOF(x, 20)), 20), state->slope - Q_CONVERT_FLOAT(1.0, 20), 20, 20, 20)), 30, 20, 30)
```
Same as the comment above, let's focus on the code part of data processing first.

johnylin76 commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,849 @@
+}
+
+/* Input: Q12.20, Q20.12 -> Output: Q20.12 */
+static int32_t slope_at(struct drc_state *state, int32_t x, int32_t k)
```
Sure, just like drc_math.h/.c which we have in CRAS.

johnylin76 replied:
Sure, just like drc_math.h/.c which we have in CRAS.

johnylin76 commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,849 @@
+		}
+
+		/* Re-calculate based on geometric mean */
+		k = Q_CONVERT_FLOAT(sqrtf(Q_CONVERT_QTOF(min_k, 12) * Q_CONVERT_QTOF(max_k, 12)), 12);
```
Same as the comment above, let's focus on the code part of data processing first.

johnylin76 replied:
Same as the comment above, let's focus on the code part of data processing first.

cujomalainey commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,849 @@
+	/* Approximate k given initial values */
+	min_k = Q_CONVERT_FLOAT(0.1, 12);
+	max_k = Q_CONVERT_FLOAT(10000.0, 12);
+	k = Q_CONVERT_FLOAT(5.0, 12);
```
@lgirdwood this is one the locations k is defined

johnylin76 commented:
I updated a compile-ok version of commit with the following modifications:
1. As the improvement from last version, hidden heavy parameter pre-calculation from component initialization. They should be calculated by CPU and passed by component config.
2. Implemented drc_func (process one source buffer and produce one sink buffer) in the manner of floating point. This is absolutely not a submit-able implementation however it could be worked first for verifying component architecture.

singalsu commented on src/audio/Kconfig:
```diff
@@ -134,6 +134,12 @@ config COMP_CROSSOVER
+	bool "Dynamic Range Compressor component"
+	default n
+	help
+	  Select for Dynamic Range Compressor component.
```
Maybe one or two sentences about what a DRC does would be useful.

singalsu commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,605 @@
+
+	state->processed = 0;
+
+	state->max_attack_compression_diff_db = INT32_MIN;
```
Probably OK, but this would be a very small value as linear :)

singalsu commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,605 @@
+
+	/* Allocate pre-delay (lookahead) buffers */
+	for (i = 0; i < channels; ++i) {
+		state->pre_delay_buffers[i] =
```
Would it be better to allocate in a single rzalloc() the combined need of buffers? Since the pool is organized as 2^N sizes there typically remains unused but consumed RAM with non-2^N sizes.

singalsu commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,605 @@
+	 * won't split a division of samples into two blocks of memory, so it is
+	 * easier to process. This may make the actual delay time slightly less
+	 * than the specified value, but the difference is less than 1ms. */
+	pre_delay_frames &= ~DRC_DIVISION_FRAMES_MASK;
```
It would be good to add to drc.h a note that DRC_DIVISION_FRAMES needs to be a 2^N number for this to work correctly.

singalsu commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,605 @@
+	cd->config = NULL;
+	cd->config_new = NULL;
+
+	if (bs) {
```
The comp_data_blob_handler_new() can do this and save code lines.

singalsu commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,605 @@
+		comp_info(dev, "drc_cmd_get_data(), SOF_CTRL_CMD_BINARY");
+
+		/* Need subtract headers to calculate payload chunk size */
+		max_size -= sizeof(struct sof_ipc_ctrl_data) +
```
Use comp_data_blob_get_cmd()

singalsu commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,605 @@
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "drc_cmd_set_data(), SOF_CTRL_CMD_BINARY");
+
+		/* Check that there is no work-in-progress previous request */
```
Use comp_data_blob_set_cmd()

singalsu commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,605 @@
+				  sink_list);
+
+	/* Check for changed configuration */
+	if (cd->config_new) {
```
Use comp_is_new_data_blob_available()

singalsu commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,605 @@
+	/* Initialize DRC */
+	comp_info(dev, "drc_prepare(), source_format=%d, sink_format=%d",
+		  cd->source_format, cd->source_format);
+	if (cd->config) {
```
Use comp_get_data_blob()

singalsu commented on src/audio/drc/drc_generic.c:
```diff
@@ -0,0 +1,435 @@
+		 * => y/x = ratio_base * x^(s - 1)
+		 * => y/x = ratio_base * e^(log(x) * (s - 1))
+		 */
+		y = ratio_base * knee_expf(logf(x) * (slope - 1));
```
I'm guessing this is already in your list of things to do but probably good idea to write own fractional version of logf() to avoid the C float math library.

singalsu commented on src/audio/drc/drc_generic.c:
```diff
@@ -0,0 +1,435 @@
+/* Update detector_average from the last input division. */
+static void drc_update_detector_average(struct drc_state *state, struct sof_drc_params *p, int nch)
+{
+	float abs_input_array[DRC_DIVISION_FRAMES];
```
Do you plan to keep the floats? It might be OK with HiFi3 already but better to check impact to code size and speed.

cujomalainey replied:
I don't think we have the hifi spec handy, would need help managing that implementation

singalsu replied:
@cujomalainey I'll help to get the spec to you. It's very useful.

singalsu commented on src/audio/drc/drc_generic.c:
```diff
@@ -0,0 +1,435 @@
+
+	for (i = 0; i < DRC_DIVISION_FRAMES; i++) {
+		/* Compute compression amount from un-delayed signal */
+		float abs_input = abs_input_array[i];
```
Variables introductions have been in function start. It's probably to see immediately how much the function uses stack.

singalsu commented on src/audio/drc/drc_generic.c:
```diff
@@ -0,0 +1,435 @@
+		state->max_attack_compression_diff_db = INT32_MIN;
+
+		/* Fix gremlins. */
+		if (isbadf(compression_diff_db))
```
If you plan to keep floats, would fast-math optimization avoid need for isnormal() check?

cujomalainey replied:
we are planning to remove floats. just doing floats initially for "get it up and running" then will move to fixed point

singalsu commented on src/audio/drc/drc_generic.c:
```diff
@@ -0,0 +1,435 @@
+				/* Warp pre-compression gain to smooth out sharp
+				 * exponential transition points.
+				 */
+				float post_warp_compressor_gain = warp_sinf(x[j] + base);
```
sinf() looks expensive computationally. Is this executed when processing copy()?

cujomalainey replied:
I think it is called from copy

johnylin76 replied:
Unfortunately all math functions in the latest code are called from copy, and will be called multiple times.
We will try to implement the approximated cheaper alternatives.

singalsu commented on src/audio/drc/drc_generic.c:
```diff
@@ -0,0 +1,435 @@
+	int i, j, ch, inc;
+	float sample;
+
+	/* Exponential approach to desired gain. */
```
Would exponential settling of a recursive filter (IIR) be more economical to compute?

singalsu commented on src/include/sof/audio/drc/drc.h:
```diff
@@ -0,0 +1,103 @@
+#define DRC_MAX_PRE_DELAY_FRAMES 1024
+#define DRC_MAX_PRE_DELAY_FRAMES_MASK (DRC_MAX_PRE_DELAY_FRAMES - 1)
+#define DRC_DEFAULT_PRE_DELAY_FRAMES 256
+#define DRC_DIVISION_FRAMES 32
```
Add comment about 2^N value need

singalsu commented on src/include/sof/audio/drc/drc_math.h:
```diff
@@ -0,0 +1,73 @@
+#define DRC_NEG_TWO_DB 0.7943282347242815f /* -2dB = 10^(-2/20) */
+
+#ifndef max
+#define max(a, b)                                                              \
```
MAX and MIN can be used from sof/math/numbers.h.

singalsu commented on src/include/sof/audio/drc/drc_math.h:
```diff
@@ -0,0 +1,73 @@
+	})
+#endif
+
+static inline float decibels_to_linear(float decibels)
```
There's a version in sof/math/decibels.h. Though I'm not sure if fit your needs. To achieve fast speed it's not as accurate expf(). The other way for lin2db is missing from there. If you plan to do a fractional version contributing there is welcome!

johnylin76 commented:
Hi @singalsu , thanks for the detailed reviews.
For C float math library usage, I think our plan is to avoid all of those (at least avoid from processing copy()). The implementation right now could help us on verifying the overall architecture by testbench in the first hand, then reducing those heavy arithmetical computations.

johnylin76 commented:
The PR is updated. I have verified the functionality of DRC implementation by testbench, comparing input and output stream samples with the golden model - DRC kernel in ChromeOS user space.

Note that this DRC firmware is still floating-point calculations. So of course the next step is to implement the fixed-point arithmetic version upon it.

cujomalainey commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,606 @@
+
+	drc = COMP_GET_IPC(dev, sof_ipc_comp_process);
+	ret = memcpy_s(drc, sizeof(*drc), ipc_drc,
+		       sizeof(struct sof_ipc_comp_process));
```
i am not sure if i am missing something but I dont see where drc get used/stored? Are we leaking here?

johnylin76 replied:
The code is performing memory copy from ipc_drc (obtained from drc_new() input) to dev->comp (drc = dev->comp)
dev->comp will be used in future process.

cujomalainey replied:
ah yes it is much clearer with the update, thanks

cujomalainey commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,606 @@
+
+	comp_info(dev, "drc_params()");
+
+	err = drc_verify_params(dev, params);
```
I don't think `drc_verify_params` needs to be its own function, all it does is call another function

cujomalainey commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,606 @@
+{
+	comp_info(dev, "drc_trigger()");
+
+	return comp_set_state(dev, cmd);
```
Should we check the processing func to make sure its set so we don't dereference a null pointer?

cujomalainey commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,606 @@
+		}
+		comp_info(dev, "drc_prepare(), DRC is configured.");
+	} else {
+		cd->drc_func = drc_find_proc_func_pass(cd->source_format);
```
we should probably use a lookup table to find the correct function

cujomalainey replied:
ignore this

cujomalainey commented on src/audio/drc/drc_generic.c:
```diff
@@ -0,0 +1,453 @@
+}
+
+/* Convert an audio sample from floating point format to s16 format and saturate to [-1.0, 1.0].*/
+static int16_t drc_float2s16(float f)
```
is this temp?

johnylin76 replied:
Yes, this is temp. We should replace all float usage.

cujomalainey commented on src/include/ipc/topology.h:
```diff
@@ -47,6 +47,7 @@ enum sof_comp_type {
 	SOF_COMP_ASRC,		/**< Asynchronous sample rate converter */
 	SOF_COMP_DCBLOCK,
 	SOF_COMP_SMART_AMP,		/**< smart amplifier component */
+	SOF_COMP_DRC,
```
i don't think we need this since we are using uuid or is this just here for testbench sake?

johnylin76 replied:
Yes, it is temp for testbench sake and should be replaced by uuid usage

lgirdwood reviewed (commented):
@johnylin76 I think all the infrastructure code looks fine here. I guess we are now just dealing with math related optimisations now for merge readiness ?
@singalsu is there any way we can test these more complex components in CI ? I guess we can do an initial testbench smoke test in CI of each individual component  (running on host PC with known input and an expectation of output) ?

singalsu commented:
@lgirdwood I'm doing at the moment a test case for volume component controls with nocodec I2S loopback topology. I think the testbench could also execute the tests on a ssh remote target. The testbench start-end locations are marked with the chirps so it can sync to capture of a real test run too.

cujomalainey commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,477 @@
+		return -ENOMEM;
+
+	memset(state->pre_delay_buffers[0], 0,
+	       sample_bytes * channels * DRC_MAX_PRE_DELAY_FRAMES);
```
this math is duplicated a couple times, lets either use sizeof or store it

cujomalainey commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,477 @@
+		       sizeof(struct sof_ipc_comp_process));
+	assert(!ret);
+
+	dev->state = COMP_STATE_INIT;
```
What is init? It is not on the state graph https://thesofproject.github.io/latest/developer_guides/firmware/components/component-overview.html#handle-the-component-device-state

cujomalainey replied:
https://github.com/thesofproject/sof-docs/issues/290

cujomalainey commented on src/audio/drc/drc_generic.c:
```diff
@@ -0,0 +1,469 @@
+		x4 = x2 * x2;
+		release_frames = kA + kB * x + kC * x2 + kD * x3 + kE * x4;
+
+#define kSpacingDb 5
```
we should add these config fields to the Kconfig and adjust their names as needed

johnylin76 commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,477 @@
+	rfree(dev);
+}
+
+static int drc_params(struct comp_dev *dev,
```
It would be used as a component API:
https://github.com/thesofproject/sof/pull/3415/files#diff-01832d594dc43038b557d1d9f6cba8d2R457

johnylin76 commented:
I have tried some fixed-point calculations and compared MSE than floating-point calculations with my test streams to make sure there is no significant approximation errors. Please see src/include/sof/audio/drc/drc_math.h

singalsu commented on src/audio/Kconfig:
```diff
@@ -134,6 +134,14 @@ config COMP_CROSSOVER
+	help
+	  Select for Dynamic Range Compressor (DRC) component. A DRC can be used
+	  to reduce the volume of loud sounds thus compressing an audio signal's
+	  dynamic range.
```
I think it can also amplify silent sounds (in addition to limit output) based on the defined compression curve. Right?

singalsu commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,483 @@
+
+	for (i = 1; i < channels; ++i) {
+		state->pre_delay_buffers[i] =
+			state->pre_delay_buffers[i - 1] + bytes_per_channel;
```
I realized that this workes because pre_delay_buffers[] is int8_t *. The data is really int16_t it seems so it's type casted when used. If the data is always int16_t maybe the definition of pre_delay_buffers[] and this pointer arithmetic should change? But if it's stream dependent varying type this is probably best.

johnylin76 replied:
Yeah, it is stream dependent varying type. Although currently I only implemented S16 case.

singalsu commented on src/audio/drc/drc.c:
```diff
@@ -0,0 +1,483 @@
+	drc_reset_state(&cd->state);
+
+	/* Allocate pre-delay buffers */
+	uint32_t sample_bytes = get_sample_bytes(cd->source_format);
```
Could this line move up to variables definitions?

lgirdwood commented:
@cujomalainey @johnylin76 should we merge the DRC once it's functionally correct and then incrementally merge the optimisations as they are ready ? I'm thinking this may reduce the technical debt that has to be carried and reduce the reviewer load ?

cujomalainey commented:
I'm ok with merging floating point if you are @lgirdwood, i think that would be a good way to keep the review clean.

lgirdwood commented:
> I'm ok with merging floating point if you are @lgirdwood, i think that would be a good way to keep the review clean.

@singalsu any comment. This would mean it would run on testbench (and would keep the review straight forward), not sure what other changes it would need to run on xtensa floating point though ?

cujomalainey commented:
My guess is little to none, but it would likely be so slow that it would overrun every system we currently support

johnylin76 commented:
@lgirdwood Currently I use a fake UUID since we haven't decide the UUID of DRC yet. Should we determine the UUID of DRC before merging?

johnylin76 commented:
I also have implemented the delay-only mode which is used when DRC is disabled. Please refer to:
https://github.com/thesofproject/sof/pull/3415/commits/abee49147cec70ebd11a2d2a21832777f976ff0a

singalsu commented:
@lgirdwood Also my +1 for merging floating point with testbench execute capability when OK for Johnny.

It reduces effort to work with optimized version. I've noticed that recent platforms (TGL) have quite a lot more processing power, even gcc built beamformer survives in real-time, so even some light optimization might allow starting real time tests with DRC.

lgirdwood commented:
> @lgirdwood Currently I use a fake UUID since we haven't decide the UUID of DRC yet. Should we determine the UUID of DRC before merging?

Yes please, just use a UUID generation tool and attach it.  We can change it when significant updates are made. Please also change PR from draft.

johnylin76 commented:
> > @lgirdwood Currently I use a fake UUID since we haven't decide the UUID of DRC yet. Should we determine the UUID of DRC before merging?
> 
> Yes please, just use a UUID generation tool and attach it. We can change it when significant updates are made. Please also change PR from draft.

Thanks. Done. Is is then for me to merge.

lgirdwood commented:
@johnylin76 merged, now good for the next phase !
